tesseract/tesseract_planning/tesseract_motion_planners/include/tesseract_motion_planners/core/utils.h

```diff
@@ -147,6 +153,582 @@ inline std::vector<Waypoint> interpolate_waypoint(const Waypoint& start, const W
+                                const CompositeInstruction& program,
+                                const tesseract_collision::ContactRequest& request =
+                                    tesseract_collision::ContactRequest(tesseract_collision::ContactTestType::FIRST),
+                                bool verbose = false)
+{
+  bool found = false;
+
+  // Flatten results
+  std::vector<std::reference_wrapper<const Instruction>> mi = flatten(program, moveFilter);
+
+  contacts.reserve(mi.size());
+  for (std::size_t iStep = 0; iStep < mi.size() - 1; ++iStep)
+  {
+    const auto* swp0 = mi.at(iStep).get().cast_const<MoveInstruction>()->getWaypoint().cast_const<StateWaypoint>();
+
+    tesseract_environment::EnvState::Ptr state = state_solver.getState(swp0->joint_names, swp0->position);
+    if (checkTrajectoryState(contacts, manager, state, request, verbose))
+    {
+      found = true;
+      if (verbose)
+      {
+        std::stringstream ss;
+        ss << "Discrete collision detected at step: " << iStep << " of " << (mi.size() - 1) << std::endl;
+
+        ss << "     Names:";
+        for (const auto& name : swp0->joint_names)
+          ss << " " << name;
+
+        ss << std::endl << "    State0: " << swp0->position << std::endl;
+
+        CONSOLE_BRIDGE_logError(ss.str().c_str());
+      }
+    }
+
+    if (found && (request.type == tesseract_collision::ContactTestType::FIRST))
+      break;
+  }
+
+  return found;
+}
+
+/**
+ * @brief Should perform a discrete collision check over the trajectory
+ * @param contacts A vector of vector of ContactMap where each indicie corrisponds to a timestep
+ * @param manager A continuous contact manager
+ * @param state_solver The environment state solver
+ * @param program The program to check for contacts
+ * @param longest_valid_segment_length Used to check collisions between two state if norm(state0-state1) >
+ * longest_valid_segment_length.
+ * @param request Contact request data
+ * @param verbose Print out found collisions
+ * @return True if collision was found, otherwise false.
+ */
+inline bool contactCheckProgram(std::vector<tesseract_collision::ContactResultMap>& contacts,
+                                tesseract_collision::DiscreteContactManager& manager,
+                                const tesseract_environment::StateSolver& state_solver,
+                                const CompositeInstruction& program,
+                                double longest_valid_segment_length,
+                                const tesseract_collision::ContactRequest& request =
+                                    tesseract_collision::ContactRequest(tesseract_collision::ContactTestType::FIRST),
+                                bool verbose = false)
+{
+  bool found = false;
+
+  // Flatten results
+  std::vector<std::reference_wrapper<const Instruction>> mi = flatten(program, moveFilter);
+
+  contacts.reserve(mi.size());
+  for (std::size_t iStep = 0; iStep < mi.size(); ++iStep)
```

Comment:
I think this needs to be `for (std::size_t iStep = 0; iStep < mi.size() - 1; ++iStep)` or else swp1 needs to just go away

Reply:
Since this is discrete it needs to handle the last state in the trajectory. There may be a cleaner way to handle this but I am doing by the logic highlighted below in the function.

```
    double dist = -1;
    if (iStep < mi.size() - 1)
      dist = (swp1->position - swp0->position).norm();

    if (dist > 0 && dist > longest_valid_segment_length)
    {
    }
    else
    {
      ... This just checks iStep for collision
    }
```

Reply:
Will line 660 not segfault here?

Reply:
I did not look into where, but I was getting an out of bounds iterator error in this function.

Reply:
You are correct. I am fixing it now.